### ISSUE TITLE
portable-net45+sl5+win8+wpa81+wp8 and xamarinmac20

### ISSUE THREAD

**Issue by tofutim** [NuGet/Home#4454]
Please read the following information before posting the issue.

## Details about Problem

NuGet product used (NuGet.exe | VS UI | Package Manager Console | dotnet.exe): Xamarin Studio

NuGet version (x.x.x.xxx): 3.5

OS version (i.e. win10 v1607 (14393.321)): OSX

Worked before? If so, with which NuGet version: 2

## Detailed repro steps so we can see the same problem

(Hoping in here from https://forums.xamarin.com/discussion/comment/249677#Comment_249677)

1. Create a new Xamarin Mac Unified project

2. Try installing Microsoft.AspNet.SignalR.Client, it will fail because 'portable-net45+sl5+win8+wpa81+wp8' is not recognized as valid for xamarinmac20.

```
Retrieving package 'Microsoft.AspNet.SignalR.Client 2.2.1' from 'nuget.org'.
For adding package 'Microsoft.AspNet.SignalR.Client.2.2.1' to project 'Eggs' that targets 'xamarinmac20'.
Install failed. Rolling back...
Package 'Microsoft.AspNet.SignalR.Client.2.2.1' does not exist in project 'Eggs'
Package 'Microsoft.AspNet.SignalR.Client.2.2.1' does not exist in folder '/Users/tim/Projects/Eggs/packages'
Executing nuget actions took 147.87 ms
Could not install package 'Microsoft.AspNet.SignalR.Client 2.2.1'. You are trying to install this package into a project that targets 'Xamarin.Mac,Version=v2.0', but the package does not contain any assembly references or content files that are compatible with that framework. For more information, contact the package author.
```

The problem is apparently in https://github.com/NuGet/NuGet.Client/blob/4cccb13833ad29d6a0bcff055460d964f1b49cfe/src/NuGet.Core/NuGet.Frameworks/DefaultPortableFrameworkMappings.cs#L125-L130 where the Xamarin framework PCL compatibility is hardcoded. Here Xamarin Mac is missing and since Xamarin Studio uses nuget 3.x, sadness. Note that if a package targets netstandard1.1 it does work.

```
                    var monoandroid = new NuGetFramework(FrameworkConstants.FrameworkIdentifiers.MonoAndroid, new Version(0, 0));
                    var monotouch = new NuGetFramework(FrameworkConstants.FrameworkIdentifiers.MonoTouch, new Version(0, 0));
                    var xamarinIOs = new NuGetFramework(FrameworkConstants.FrameworkIdentifiers.XamarinIOs, new Version(0, 0));
                    var xamarinTVOS = new NuGetFramework(FrameworkConstants.FrameworkIdentifiers.XamarinTVOS, new Version(0, 0));
                    var xamarinWatchOS = new NuGetFramework(FrameworkConstants.FrameworkIdentifiers.XamarinWatchOS, new Version(0, 0));
                    var monoFrameworks = new NuGetFramework[] { monoandroid, monotouch, xamarinIOs, xamarinWatchOS, xamarinTVOS };
```

**Comment by mrward** [comment]
We should add the Xamarin.Mac framework to the optional framework list in DefaultPortableFrameworkMappings since we want to support adding PCLs into Xamarin.Mac mobile projects.

/cc @chamons

**Comment by emgarten** [comment]
@mrward would you be able to create a PR with the correct mappings?

> Note that if a package targets netstandard1.1 it does work.

Long term netstandard will fix this. I see that the package contains netcore50 which may have been an attempt at this before the netstandard TFM was available.

**Comment by tofutim** [comment]
Looks like in https://github.com/NuGet/NuGet.Client/blob/4cccb13833ad29d6a0bcff055460d964f1b49cfe/src/NuGet.Core/NuGet.Frameworks/DefaultFrameworkMappings.cs there is already a FrameworkIdentifier for Mac
```
new KeyValuePair<string, string>(FrameworkConstants.FrameworkIdentifiers.XamarinMac, "xamarinmac"),
```
Not sure if it has to be xamarinmac20 or not.

**Comment by mrward** [comment]
I have opened a [PR](https://github.com/NuGet/NuGet.Client/issues/1161) with the changes. I tested the new NuGet binaries with Xamarin Studio locally and this fixes the problem with Xamarin.Mac projects not allowing PCLs to be installed from NuGet packages.

**Comment by tofutim** [comment]
Oh you are awesome. I have a next nuget question for you.

The SignalR portable library does not include WebSockets.
WebSockets will be available in netstandard2.0
Xamarin iOS/Mac/tvOS/watchOS/Android support WebSockets, but it seems I
would have to make a separate dll for each, is that the case? I also saw an
Apple PCL but this does not seem fully baked yet and would exclude Android.

Wish there was a xamarinstandardXX

**Comment by mrward** [comment]
@tofutim - Yes if you cannot implement it in .NETStandard v1 nor in a PCL then you would need to have an assembly for each Xamarin platform you want to support.

**Comment by rrelyea** [comment]
set to rtm so we can triage and get approval for check into 4.0.0 if possible, otherwise will go in 4.0.1.

**Comment by joelverhagen** [comment]
@mrward, how does this relate to https://github.com/NuGet/Home/issues/2662? I put out a PR (https://github.com/NuGet/NuGet.Client/pull/562) for that issue quite a while that never went in due to a rabbit-hole of discussions over email.

**Comment by joelverhagen** [comment]
The remaining Xamarin TFMs not in that list are:

```
MonoMac
XamarinPlayStation3
XamarinPlayStation4
XamarinPlayStationVita
XamarinXbox360
XamarinXboxOne
```

Does this seem right?

**Comment by mrward** [comment]
@joelverhagen - For #2662 the only impact would be just that we would want the XamarinMacNet target framework to allow PCLs to be installed. That PR is still valid.

MonoMac has never been used and there are no plans to use that target framework. The other target frameworks in that list are not currently being used as far as I am aware. They could be added to the list. Right now only Xamarin.Mac is in active use so it was the only one I added to the list.

**Comment by tofutim** [comment]
matt, is it straight forward to install the fixed nuget binaries into
xamarin studio? or do i have to wait for new version of xamarin studio?

**Comment by mrward** [comment]
It should be straight to install fixed NuGet binaries into Xamarin Studio. The more difficult problem is building them. Which version of Xamarin Studio are you using will determine what version of NuGet needs building. You will also need to build the NuGet binaries on Windows. If you let me know what version of Xamarin Studio you are using I can build some patched binaries for you.

**Comment by rrelyea** [comment]
@mrward - this is difficult to do in v4, but 4.0.1 would be easier. if you are building custom binaries anyway for xamarin studio use, is 4.0.0 vs 4.0.1 (dev15 vs 15.1) critical for you?

**Comment by mrward** [comment]
@rrelyea - No problem with this not being included in 4.0.0. I can build custom binaries for Xamarin Studio if anyone needs them.

**Comment by tofutim** [comment]
Xamarin Studio Business
Version 6.2 (build 1798)
Installation UUID: db3cdb23-e5f7-4e4e-913c-11179c7be9fa
Runtime:
Mono 4.8.0 (mono-4.8.0-branch/084f912) (64-bit)
GTK+ 2.24.23 (Raleigh theme)

Package version: 408000478


I jumped to 6.2 (Beta) from 6.1.5 (Stable) to check if it would work. 6.2
would be sufficient.

**Comment by mrward** [comment]
Patched NuGet binaries with this fix for Xamarin Studio 6.2 are available on [GitHub](https://github.com/mrward/nuget-binary/tree/3.5.0-xamarinmac-pcl-support).

**Comment by emgarten** [comment]
Fixed with NuGet/NuGet.Client@9bb3952a37ac26664d11079fea5d2b7b352264ed
